pulsar-client-cpp/lib/ProducerImpl.cc

```diff
@@ -500,9 +506,26 @@ void ProducerImpl::sendAsyncWithStatsUpdate(const Message& msg, const SendCallba
                 return;
             }
 
-            sendMessage(OpSendMsg{msgMetadata, encryptedPayload,
-                                  (chunkId == totalChunks - 1) ? callback : nullptr, producerId_, sequenceId,
-                                  conf_.getSendTimeout(), 1, uncompressedSize});
+            OpSendMsg op =
+                OpSendMsg{msgMetadata, encryptedPayload, (chunkId == totalChunks - 1) ? callback : nullptr,
+                          producerId_, sequenceId,       conf_.getSendTimeout(),
+                          1,           uncompressedSize};
+
+            if (!chunkingEnabled_) {
+                const uint32_t msgMetadataSize = op.metadata_.ByteSize();
+                const uint32_t payloadSize = op.payload_.readableBytes();
+                const uint32_t msgHeadersAndPayloadSize = msgMetadataSize + payloadSize;
+                if (msgHeadersAndPayloadSize > maxMessageSize) {
```

Comment:
```suggestion
                if (msgHeadersAndPayloadSize > maxMessageSize) {
                    lock.unlock();
```

We can unlock it since the lock is acquired only for `sendMessage`.